server/src/main/java/org/elasticsearch/action/admin/indices/rollover/MetadataRolloverService.java

```diff
@@ -190,6 +204,20 @@ private static NameResolution resolveDataStreamRolloverNames(Metadata metadata,
         return new NameResolution(originalWriteIndex.getIndex().getName(), null, dataStream.nextWriteIndexAndGeneration(metadata).v1());
     }
 
+    private static NameResolution resolveDataStreamFailureStoreRolloverNames(Metadata metadata, DataStream dataStream) {
+        /*
+         * TODO: Do we need to check if this is null?
+         *  By this point, we have already validated that the data stream has to have a failure store.
+         *  (but Intellij is warning that `dataStream.getFailureStoreWriteIndex()` might be null)
```

Comment:
@jbaiera interesting one, from a first glance, I disagreed on differentiating between a conditional and an unconditional rollover. But with some more thinking, I think we do not really change the behaviour, we just default all conditions to false.

The user does not know that we are lazily creating the failure store backing indices so they shouldn't have to worry about it. So, I think your proposal is the one that makes more sense:
- Explicit unconditional rollover: creates the first index
- Conditional rollover: fails as no conditions are met, since we have no data, no creation date to calculate the age etc etc..
+1

The only concern would be can a user shoot themselves in the foot by creating without realising an index. I can't see anything going wrong, because from the creation of the index we are going back to business as usual, but I wanted to leave it out there in case someone else thinks of something.

Reply:
Hmm I definitely agree that rollover with conditions on a missing failure store index should indeed not do anything. I'm still thinking about the first scenario - a rollover without conditions. I'm trying to see what the benefit and/or use case of that scenario would be. If we were to implement it the way you both suggest, it would require additional logic/complexity, so I'm trying to assess whether that's worth it. The only use case I can come up with, is that a user enabled failure store on a DS, but want to see what the failure store index looks like exactly (e.g. mappings, settings, naming,  w/e), so they want to forcibly create that index. Do either of you foresee any other use cases? If this would be the only use case, I'm not sure I'd consider it worth the complexity.

Reply:
To answer the main inquiry here: I think considering that changing this is an increase in complexity for a case that does not yet exist (we're not lazily creating failure stores as of now) then we can probably safely rely on the assertion here and the validation elsewhere and then address this problem when the lazy instantiation work is in progress.

> The only concern would be can a user shoot themselves in the foot by creating without realising an index

I don't think it's any worse than a user running an unconditional rollover on a data stream with an empty write index. We contribute some more load on the cluster state perhaps, if things run smoothly forever then perhaps the failure store index sits around with nothing in it.

> Do either of you foresee any other use cases?

We support lazy rollover of a data stream when writing to it. We eventually want to support lazy creation of failure stores when writing to them. It could be an elegant solution if a new data stream has no failure store indices and is already marked to lazily rollover the failure store, thus using the lazy rollover operation to create the failure store on its first use. We'd need to check lazy rollover for failure stores at every use already, so instead of testing for both cases (Lazy rollover for failure store enabled? Failure store empty?) we could just test for the lazy rollover case and get the second case would be a freebie.

Reply:
I also see it from user experience perspective. The lazy creation of the failure store index is an implementation detail that should not influence how a user is experience things. For example:
- when something fails it will create the failure store index and index it there, 
- when the user is asking to rollover with conditions, we can say the conditions were evaluated to false, so no rollover is happening, 
- when the user is requesting an unconditional rollover, then we create a new index like a rollover would have done.

Failing an unconditional rollover here for example, would be changing the user witnessed behaviour of the rollover when we introduce the lazy failure store creation.

On the other hand I understand that the rollover is kind of an expert API so we might not be so strict here. In any case, I do not have very strong feelings about this yet, just explaining the why I am leaning towards the initial proposal.

Reply:
I'm definitely on board with lazily creating and rolling over failure store indices. I was just saying I didn't really get the benefit of supporting the use case of a _user_ rolling over an empty failure store and thus initializing it. I agree that this is a win-win when it comes to lazily creating failure stores and lazily rolling them over. As @jbaiera said, let's do this when we're actually going to implement lazily rollover for failure stores, and stick to the assertion for now.